### ISSUE TITLE
Virtualized Scrolling

### ISSUE THREAD

**Issue by msssk** [dojo/dgrid#40]
## Enhancement

The default implementation of a scrolling container leaves rendering of the scrollbar and positioning of the scrolled content to the browser. This requires creating a content node of sufficient dimensions to contain all data (whether the data is currently visible or not, or rendered or not). For extremely large numbers of rows in a grid, this can cause erroneous behavior in browsers.

Virtualized scrolling provides a scrollbar that is not directly connected to any content. The virtual scrollbar should be able to synchronize with a separate node containing scrollable content. If the content node is scrolled it should calculate its own position as a percentage value and set the virtual scrollbar's scroll position to match. If the virtual scrollbar is scrolled it should calculate its own position as a percentage value and set the content node's scroll position to match.

### Proof of Concept

A standalone POC has been developed that has acceptable performance in current desktop browsers (IE 11, Edge, Chrome, Firefox).

https://github.com/msssk/dgrid/blob/faux-scroller/test/scroll.html

**Comment by msssk** [comment]
A dgrid implementation is underway:
https://github.com/msssk/dgrid2/tree/scrolling

I am seeing unacceptable performance degradation in IE, Edge, and Firefox (Chrome works perfectly).
I am working on a direct transplant of the POC code into dgrid to see if performance issues persist:
https://github.com/msssk/dgrid2/tree/scrolling-poc

**Comment by msssk** [comment]
The [scrolling branch](https://github.com/msssk/dgrid2/tree/scrolling) has a working implementation within dgrid. It is based on the [POC](https://github.com/msssk/dgrid2/tree/scrolling-poc) and there's a few lessons learned to consider moving forward:

- optimization attempts have been counterproductive: `requestAnimationFrame` and debouncing or throttling the scroll event handlers have resulted in worse performance. Calling `setTimeout` within a scroll event handler causes very janky scrolling in Firefox and IE.
- the current implementation uses a counter variable to prevent an infinite loop of scroll event handlers between the two scrolling nodes. Currently it seems to be working reliably (IE 11, Edge, Chrome, Firefox). In the past I attempted a simple boolean flag, but observed that there was not the expected 1-to-1 relationship between setting `node.scrollTop` and a corresponding `scroll` event being emitted by `node`. Because of this unreliability I moved from a boolean flag to a counter variable. Even with that change I observed discrepancies between how many times `node.scrollTop` was set and how many times the `scroll` event was emitted by `node`. Considering these past experiences I am cautiously optimistic about the current implementation. It seems to be working well, but there may be problems lurking under the surface.
